In [1]:
from urllib import request
import pandas as  pd
import numpy as np


#### <span style="color:blue"> 1. Download do Dataset do Portal de Dados Abertos </span>

In [2]:
file_url = 'https://ons-dl-prod-opendata.s3.amazonaws.com/dataset/reservatorio/RESERVATORIOS.csv'
source_file = 'RESERVATORIOS.csv'
request.urlretrieve(file_url , source_file )

('RESERVATORIOS.csv', <http.client.HTTPMessage at 0x17624764070>)

#### <span style="color:blue"> 2. Importar arquivo </span>

In [3]:
dataframe = pd.read_csv(source_file, sep=';',encoding = "utf-8")

if (dataframe.empty) :
    print('DATAFRAME VAZIO')
else:
  print('shape:', dataframe.shape)
  print('columns:', dataframe.columns)

shape: (158, 21)
columns: Index(['nom_reservatorio', 'tip_reservatorio', 'cod_resplanejamento',
       'cod_posto', 'nom_usina', 'ceg', 'id_subsistema', 'nom_subsistema',
       'nom_bacia', 'nom_rio', 'nom_ree', 'dat_entrada', 'val_cotamaxima',
       'val_cotaminima', 'val_volmax', 'val_volmin', 'val_volutiltot',
       'val_produtibilidadeespecifica', 'val_produtividade65volutil',
       'val_tipoperda', 'val_perda'],
      dtype='object')


In [4]:
dataframe.head(3)

nom_reservatorio        tip_reservatorio  cod_resplanejamento  cod_posto  \
0      14 DE JULHO               FIO DAGUA                 99.0      284.0   
1      A. VERMELHA  RESERVATÓRIO COM USINA                 18.0       18.0   
2          AIMORES               FIO DAGUA                143.0      148.0   

       nom_usina                    ceg id_subsistema        nom_subsistema  \
0    14 DE JULHO  UHE.PH.RS.000012-4.01             S                   SUL   
1  ÁGUA VERMELHA  UHE.PH.MG.000041-8.01            SE  SUDESTE/CENTRO-OESTE   
2        AIMORÉS  UHE.PH.MG.000042-6.01            SE  SUDESTE/CENTRO-OESTE   

  nom_bacia        nom_rio  ... dat_entrada val_cotamaxima  val_cotaminima  \
0     JACUI  TAQUARI-ANTAS  ...  2008-12-25          104.0           103.0   
1    GRANDE         GRANDE  ...  1978-03-01          383.3           373.3   
2      DOCE           DOCE  ...  2005-07-30           90.0            89.6   

   val_volmax  val_volmin  val_volutiltot  val_produtibilidadeespecifica  \
0       55.18       50.47            4.71                       0.008927   
1    11025.00     5856.00         5169.22                       0.008825   
2      185.52      156.55           12.14                       0.008926   

   val_produtividade65volutil  val_tipoperda val_perda  
0                      0.3016              m      0.74  
1                      0.4625              m      0.75  
2                      0.2558              m      0.32  

[3 rows x 21 columns]

#### <span style="color:blue"> 3.Explorando o conteúdo do arquivo </span>

##### <span style="color:#000080"> 3.1 Quantidade de valores não nulos </span>

In [5]:
dataframe.count()

nom_reservatorio                 158
tip_reservatorio                 158
cod_resplanejamento              156
cod_posto                        156
nom_usina                        158
ceg                              158
id_subsistema                    158
nom_subsistema                   158
nom_bacia                        158
nom_rio                          158
nom_ree                          157
dat_entrada                      158
val_cotamaxima                   158
val_cotaminima                   158
val_volmax                       158
val_volmin                       158
val_volutiltot                   158
val_produtibilidadeespecifica    157
val_produtividade65volutil       157
val_tipoperda                    156
val_perda                        156
dtype: int64

##### <span style="color:#000080"> 3.2 Quantidade de valores nulos </span>

In [6]:
dataframe.isnull().sum()

nom_reservatorio                 0
tip_reservatorio                 0
cod_resplanejamento              2
cod_posto                        2
nom_usina                        0
ceg                              0
id_subsistema                    0
nom_subsistema                   0
nom_bacia                        0
nom_rio                          0
nom_ree                          1
dat_entrada                      0
val_cotamaxima                   0
val_cotaminima                   0
val_volmax                       0
val_volmin                       0
val_volutiltot                   0
val_produtibilidadeespecifica    1
val_produtividade65volutil       1
val_tipoperda                    2
val_perda                        2
dtype: int64

#### <span style="color:blue"> 4. Estatísticas </span>

In [7]:
# Variaveis do tipo "TEXTOS"
dataframe.describe(include=[object])  

nom_reservatorio tip_reservatorio    nom_usina                    ceg  \
count               158              158          158                    158   
unique              154                3          158                    154   
top     P. AFONSO 1,2,3        FIO DAGUA  14 DE JULHO  UHE.PH.PR.001161-4.01   
freq                  3               95            1                      2   

       id_subsistema        nom_subsistema  nom_bacia nom_rio nom_ree  \
count            158                   158        158     158     157   
unique             4                     4         23      60      12   
top               SE  SUDESTE/CENTRO-OESTE  PARANAIBA  GRANDE  PARANA   
freq             106                   106         19      12      44   

       dat_entrada val_tipoperda  
count          158           156  
unique         150             2  
top     1990-12-31             m  
freq             3           152

In [8]:
# Variaveis do tipo "NUMERO"
dataframe.describe(include=[np.number])
 
# ou dataframe.describe(exclude=[object]) 

cod_resplanejamento   cod_posto  val_cotamaxima  val_cotaminima  \
count           156.000000  156.000000      158.000000      158.000000   
mean            133.307692  163.230769      401.514304      394.356203   
std              93.231492   95.085350      224.098751      220.282832   
min               1.000000    1.000000        0.000000        0.000000   
25%              49.750000   80.250000      251.625000      246.000000   
50%             119.500000  173.500000      371.715000      370.465000   
75%             197.750000  249.500000      567.250000      558.650000   
max             315.000000  318.000000      913.000000      899.000000   

         val_volmax    val_volmin  val_volutiltot  \
count    158.000000    158.000000      158.000000   
mean    3824.309095   1942.608494     2142.582880   
std     8170.982305   3984.833585     5988.146545   
min        0.000000      0.000000        0.000000   
25%       93.782500     59.750000       12.320000   
50%      648.500000    444.175000      137.990000   
75%     3350.500000   2207.525000     1048.082500   
max    54400.000000  27695.190000    43250.000000   

       val_produtibilidadeespecifica  val_produtividade65volutil   val_perda  
count                     157.000000                  157.000000  156.000000  
mean                        0.008698                    0.671296    1.568013  
std                         0.001234                    0.876796    2.732385  
min                         0.000000                    0.000000    0.000000  
25%                         0.008751                    0.240800    0.377500  
50%                         0.008916                    0.402500    0.745000  
75%                         0.009010                    0.803200    1.605000  
max                         0.009218                    6.354200   23.780000

In [9]:
# Estatistica sobre uma das colunas selecionadas
dataframe.val_volutiltot.describe()

count      158.000000
mean      2142.582880
std       5988.146545
min          0.000000
25%         12.320000
50%        137.990000
75%       1048.082500
max      43250.000000
Name: val_volutiltot, dtype: float64

#### <span style="color:blue"> 5. Análise sobre os resultados - Foco em NOME DO RESERVATÓRIO</span>

<i> A função <b>value_counts()</b> do pandas é valiosíssima nas análises de dados, o que ela faz é verificar a contagem de dados iguais em um dataframe ou em uma coluna, e retorna uma Series com esse resultado.</i>  

In [10]:
dataframe.nom_reservatorio.value_counts(ascending=False)

P. AFONSO 1,2,3    3
HENRY BORDEN       2
ITAIPU             2
14 DE JULHO        1
PASSO FUNDO        1
                  ..
G. P. SOUZA        1
GARIBALDI          1
GUAPORE            1
GUILM. AMORIM      1
XINGO              1
Name: nom_reservatorio, Length: 154, dtype: int64

<i>A função <b>duplicated()</b> nos retorna uma série booleana denotando linhas duplicadas. Os resultados são Verdadeiro(True) e Falso(False), sendo verdadeiros os valores duplicados.</i>

In [11]:
resultado = dataframe.duplicated(subset=['nom_reservatorio'], keep=False)
dataframe[resultado]

nom_reservatorio tip_reservatorio  cod_resplanejamento  cod_posto  \
55     HENRY BORDEN        FIO DAGUA                119.0      318.0   
56     HENRY BORDEN        FIO DAGUA                119.0      318.0   
63           ITAIPU        FIO DAGUA                 66.0      266.0   
64           ITAIPU        FIO DAGUA                 66.0      266.0   
96  P. AFONSO 1,2,3        FIO DAGUA                174.0      174.0   
97  P. AFONSO 1,2,3        FIO DAGUA                174.0      174.0   
98  P. AFONSO 1,2,3        FIO DAGUA                174.0      174.0   

            nom_usina                    ceg id_subsistema  \
55  HENRY BORDEN EXT.  UHE.PH.SP.001084-7.01            SE   
56  HENRY BORDEN SUB.  UHE.PH.SP.001084-7.01            SE   
63       ITAIPU 60 HZ  UHE.PH.PR.001161-4.01            SE   
64       ITAIPU 50 HZ  UHE.PH.PR.001161-4.01            SE   
96    PAULO AFONSO II  UHE.PH.BA.027048-2.01            NE   
97   PAULO AFONSO III  UHE.PH.BA.027049-0.01            NE   
98     PAULO AFONSO I  UHE.PH.BA.002012-5.01            NE   

          nom_subsistema      nom_bacia        nom_rio  ... dat_entrada  \
55  SUDESTE/CENTRO-OESTE          TIETE        CUBATÃO  ...  1926-04-01   
56  SUDESTE/CENTRO-OESTE          TIETE        CUBATÃO  ...  1956-01-01   
63  SUDESTE/CENTRO-OESTE         PARANA         PARANÁ  ...  1987-01-01   
64  SUDESTE/CENTRO-OESTE         PARANA         PARANÁ  ...  1984-02-01   
96              NORDESTE  SAO FRANCISCO  SÃO FRANCISCO  ...  1961-04-30   
97              NORDESTE  SAO FRANCISCO  SÃO FRANCISCO  ...  1971-04-30   
98              NORDESTE  SAO FRANCISCO  SÃO FRANCISCO  ...  1955-01-31   

   val_cotamaxima  val_cotaminima  val_volmax  val_volmin  val_volutiltot  \
55         728.75          728.75        1.00        1.00             1.0   
56         728.75          728.75        1.00        1.00             1.0   
63         220.30          219.00    29403.91    27695.19         19000.0   
64         220.30          219.00    29403.91    27695.19         19000.0   
96         230.30          228.30       26.00       26.00             9.8   
97         230.30          228.30       26.00       26.00             9.8   
98         230.30          228.30       26.00       26.00             9.8   

    val_produtibilidadeespecifica  val_produtividade65volutil  val_tipoperda  \
55                       0.007939                      5.6808              m   
56                       0.007939                      5.6808              m   
63                       0.009037                      1.0244              m   
64                       0.009037                      1.0244              m   
96                       0.008792                      0.3800              %   
97                       0.008792                      0.3800              %   
98                       0.008792                      0.3800              %   

   val_perda  
55     10.67  
56     10.67  
63      1.33  
64      1.33  
96      1.08  
97      1.08  
98      1.08  

[7 rows x 21 columns]

#### <span style="color:blue"> 6. Conclusão</span>

- Não temos linhas repetidas no Dataset
- Observamos que o resultado quantitativo da estatística para "nom_reservatorio" teve count=158 e unique=154 sugerindo repetição.
- Numa investigação específica sobre Nome de Reservatórios, encontramos 3 nomes iguais POREM para para USINAS DIFERENTES.